File: Base de datos/convert_copy_to_insert.py
#!/usr/bin/env python3
"""
Convierte un dump de PostgreSQL con formato COPY FROM stdin a INSERT statements
"""

import re
import sys

def convert_copy_to_insert(input_file, output_file):
    with open(input_file, 'r', encoding='utf-8') as f:
        content = f.read()
    
    # Patrón para encontrar bloques COPY
    pattern = r'COPY ([^\(]+)\s*\(([^)]+)\)\s+FROM\s+stdin;(.*?)\\\\\.', re.DOTALL
    
    def replace_copy(match):
        table_name = match.group(1).strip()
        columns = match.group(2).strip()
        data_block = match.group(3).strip()
        
        if not data_block:
            return f"-- No data for {table_name}\n"
        
        lines = data_block.split('\n')
        insert_statements = []
        
        for line in lines:
            line = line.strip()
            if not line:
                continue
                
            # Separar por tabs
            values = line.split('\t')
            
            # Escapar valores
            escaped_values = []
            for val in values:
                if val == '\\N' or val == '':
                    escaped_values.append('NULL')
                else:
                    # Escapar comillas simples
                    val = val.replace("'", "''")
                    escaped_values.append(f"'{val}'")
            
            values_str = ', '.join(escaped_values)
            insert_statements.append(f"INSERT INTO {table_name} ({columns}) VALUES ({values_str});")
        
        return '\n'.join(insert_statements) + '\n'
    
    # Reemplazar todos los bloques COPY
    converted = re.sub(pattern, replace_copy, content)
    
    with open(output_file, 'w', encoding='utf-8') as f:
        f.write(converted)
    
    print(f"✅ Conversión completada: {output_file}")

if __name__ == '__main__':
    input_file = r'c:\laragon\www\grupo_10_DigitalGames\Base de datos\data.sql'
    output_file = r'c:\laragon\www\grupo_10_DigitalGames\Base de datos\data_insert_format.sql'
    convert_copy_to_insert(input_file, output_file)

File: Base de datos/convert_copy_to_insert.js
const fs = require('fs');
const path = require('path');

const inputFile = path.join(__dirname, 'data.sql');
const outputFile = path.join(__dirname, 'data_insert_format.sql');

console.log('Leyendo archivo:', inputFile);
const content = fs.readFileSync(inputFile, 'utf8');

// Patrón para encontrar bloques COPY
const copyPattern = /COPY ([^\(]+)\s*\(([^)]+)\)\s+FROM\s+stdin;([\s\S]*?)\\\\\./gm;

let converted = content;
let match;
let replacements = [];

// Encontrar todos los matches primero
copyPattern.lastIndex = 0;
while ((match = copyPattern.exec(content)) !== null) {
    const tableName = match[1].trim();
    const columns = match[2].trim();
    const dataBlock = match[3].trim();

    if (!dataBlock) {
        replacements.push({
            original: match[0],
            replacement: `-- No data for ${tableName}\n`
        });
        continue;
    }

    const lines = dataBlock.split('\n');
    const insertStatements = [];

    for (const line of lines) {
        const trimmedLine = line.trim();
        if (!trimmedLine) continue;

        // Separar por tabs
        const values = trimmedLine.split('\t');

        // Escapar valores
        const escapedValues = values.map(val => {
            if (val === '\\N' || val === '') {
                return 'NULL';
            } else {
                // Escapar comillas simples
                const escaped = val.replace(/'/g, "''");
                return `'${escaped}'`;
            }
        });

        const valuesStr = escapedValues.join(', ');
        insertStatements.push(`INSERT INTO ${tableName} (${columns}) VALUES (${valuesStr});`);
    }

    replacements.push({
        original: match[0],
        replacement: insertStatements.join('\n') + '\n'
    });
}

// Aplicar reemplazos
for (const { original, replacement } of replacements) {
    converted = converted.replace(original, replacement);
}

fs.writeFileSync(outputFile, converted, 'utf8');
console.log('✅ Conversión completada:', outputFile);
console.log('📝 Total de bloques COPY convertidos:', replacements.length);
